nixos/modules/services/web-servers/traefik.nix

```diff
@@ -4,56 +4,98 @@ with lib;
 
 let
   cfg = config.services.traefik;
-  configFile =
-    if cfg.configFile == null then
+  dynamicConfigFile =
+    if cfg.dynamicConfigFile == null then
       pkgs.runCommand "config.toml" {
         buildInputs = [ pkgs.remarshal ];
         preferLocalBuild = true;
       } ''
         remarshal -if json -of toml \
-          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
+          < ${pkgs.writeText "dynamic_config.json" (builtins.toJSON cfg.dynamicConfigOptions)} \
           > $out
       ''
-    else cfg.configFile;
+    else cfg.dynamicConfigFile;
+  recursiveMerge = attrList:
+    let f = attrPath:
+      zipAttrsWith (n: values:
+        if tail values == []
+        then head values
+        else
+          if all isList values
+          then unique (concatLists values)
+          else
+            if all isAttrs values
+            then f (attrPath ++ [n]) values
+            else last values
+      );
+    in f [] attrList;
+  staticConfigFile =
+    if cfg.staticConfigFile == null then
+      pkgs.runCommand "config.toml" {
+        buildInputs = [ pkgs.yj ];
+        preferLocalBuild = true;
+      } ''
+        yj -jt -i \
+          < ${pkgs.writeText "static_config.json" (builtins.toJSON (recursiveMerge [cfg.staticConfigOptions { providers.file.filename = "${dynamicConfigFile}";}]))} \
+          > $out
+      ''
+    else cfg.staticConfigFile;
 
 in {
   options.services.traefik = {
     enable = mkEnableOption "Traefik web server";
 
-    configFile = mkOption {
+    staticConfigFile = mkOption {
       default = null;
-      example = literalExample "/path/to/config.toml";
+      example = literalExample "/path/to/static_config.toml";
       type = types.nullOr types.path;
       description = ''
-        Path to verbatim traefik.toml to use.
-        (Using that option has precedence over <literal>configOptions</literal>)
+        Path to traefik's static configuration to use.
+        (Using that option has precedence over <literal>staticConfigOptions</literal> and <literal>dynamicConfigOptions</literal>)
       '';
     };
 
-    configOptions = mkOption {
+    staticConfigOptions = mkOption {
       description = ''
-        Config for Traefik.
+        Static configuration for Traefik.
       '';
       type = types.attrs;
       default = {
-        defaultEntryPoints = ["http"];
         entryPoints.http.address = ":80";
```

Comment:
Don't you need to set `entryPoints.web.address` like in the example case?

Reply:
To be honest, I didn't see anything `defaultEntryPoints` on the [official documentation site](https://docs.traefik.io/), so I skipped this.

This [issue](https://github.com/containous/traefik/issues/5184) on the official traefik GitHub repo supports this.